README.md
# Reactive Relational Database Connectivity MySQL Implementation
![Maven Central](https://img.shields.io/maven-central/v/io.asyncer/r2dbc-mysql)
![LICENSE](https://img.shields.io/github/license/asyncer-io/r2dbc-mysql)

This project contains the [MySQL][m] implementation of the [R2DBC SPI](https://github.com/r2dbc/r2dbc-spi).
This implementation is not intended to be used directly, but rather to be
used as the backing implementation for a humane client library to
delegate to. See [R2DBC Homepage](https://r2dbc.io).

This driver provides the following features:

- [x] Unix domain socket.
- [x] Execution of simple or batch statements without bindings.
- [x] Execution of prepared statements with bindings.
- [x] Reactive LOB types (e.g. BLOB, CLOB)
- [x] All charsets from MySQL, like `utf8mb4_0900_ai_ci`, `latin1_general_ci`, `utf32_unicode_520_ci`, etc.
- [x] All authentication types for MySQL, like `caching_sha2_password`, `mysql_native_password`, etc.
- [x] General exceptions with error code and standard SQL state mappings.
- [x] Secure connection with verification (SSL/TLS), auto-select TLS version for community and enterprise editions.
- [x] SSL tunnel for proxy protocol of MySQL.
- [x] Transactions with savepoint.
- [x] Native ping command that can be verifying when argument is `ValidationDepth.REMOTE`
- [x] Extensible, e.g. extend built-in `Codec`(s).

## Maintainer

This project is currently being maintained by [@jchrys](https://github.com/jchrys), since the previous owner has been inactive. We are committed to keeping this project up-to-date and improving it in collaboration with the community.

## Version compatibility / Integration tests states
[![MySQL 5.5 status](https://github.com/asyncer-io/r2dbc-mysql/workflows/MySQL%205.5/badge.svg)](https://github.com/asyncer-io/r2dbc-mysql/actions?query=workflow%3A%22MySQL+5.5%22)
[![MySQL 5.6 status](https://github.com/asyncer-io/r2dbc-mysql/workflows/MySQL%205.6/badge.svg)](https://github.com/asyncer-io/r2dbc-mysql/actions?query=workflow%3A%22MySQL+5.6%22)
[![MySQL 5.7 status](https://github.com/asyncer-io/r2dbc-mysql/workflows/MySQL%205.7/badge.svg)](https://github.com/asyncer-io/r2dbc-mysql/actions?query=workflow%3A%22MySQL+5.7%22)
[![MySQL 8.0 status](https://github.com/asyncer-io/r2dbc-mysql/workflows/MySQL%208.0/badge.svg)](https://github.com/asyncer-io/r2dbc-mysql/actions?query=workflow%3A%22MySQL+8.0%22)

In fact, it supports lower versions, in the theory, such as 4.1, 4.0, etc.

However, Docker-certified images do not have these versions lower than 5.5.0, so tests are not integrated on these versions.

## Maven

```xml

<dependency>
  <groupId>io.asyncer</groupId>
  <artifactId>r2dbc-mysql</artifactId>
  <version>0.9.0</version>
</dependency>
```

## Gradle

### Groovy DSL

```groovy
dependencies {
    implementation 'io.asyncer:r2dbc-mysql:0.9.0'
}
```

### Kotlin DSL

```kotlin
dependencies {
    // Maybe should to use `compile` instead of `implementation` on the lower version of Gradle.
    implementation("io.asyncer:r2dbc-mysql:0.9.0")
}
```

## Getting Started

Here is a quick teaser of how to use R2DBC MySQL in Java:

### URL Connection Factory Discovery

```java
// Notice: the query string must be URL encoded
ConnectionFactory connectionFactory = ConnectionFactories.get(
    "r2dbcs:mysql://root:database-password-in-here@127.0.0.1:3306/r2dbc?" +
    "zeroDate=use_round&" +
    "sslMode=verify_identity&" +
    "useServerPrepareStatement=true&" +
    "tlsVersion=TLSv1.3%2CTLSv1.2%2CTLSv1.1&" +
    "sslCa=%2Fpath%2Fto%2Fmysql%2Fca.pem&" +
    "sslKey=%2Fpath%2Fto%2Fmysql%2Fclient-key.pem&" +
    "sslCert=%2Fpath%2Fto%2Fmysql%2Fclient-cert.pem&" +
    "sslKeyPassword=key-pem-password-in-here"
)

// Creating a Mono using Project Reactor
Mono<Connection> connectionMono = Mono.from(connectionFactory.create());
```

> It is just example, see also Programmatic Connection Factory Discovery for more options.

Or use unix domain socket like following:

```java
// Minimum configuration for unix domain socket
ConnectionFactory connectionFactory = ConnectionFactories.get("r2dbc:mysql://root@unix?unixSocket=%2Fpath%2Fto%2Fmysql.sock")

Mono<Connection> connectionMono = Mono.from(connectionFactory.create());
```

### Programmatic Connection Factory Discovery

```java
ConnectionFactoryOptions options = ConnectionFactoryOptions.builder()
    .option(DRIVER, "mysql")
    .option(HOST, "127.0.0.1")
    .option(USER, "root")
    .option(PORT, 3306)  // optional, default 3306
    .option(PASSWORD, "database-password-in-here") // optional, default null, null means has no password
    .option(DATABASE, "r2dbc") // optional, default null, null means not specifying the database
    .option(CONNECT_TIMEOUT, Duration.ofSeconds(3)) // optional, default null, null means no timeout
    .option(Option.valueOf("socketTimeout"), Duration.ofSeconds(4)) // optional, default null, null means no timeout
    .option(SSL, true) // optional, default sslMode is "preferred", it will be ignore if sslMode is set
    .option(Option.valueOf("sslMode"), "verify_identity") // optional, default "preferred"
    .option(Option.valueOf("sslCa"), "/path/to/mysql/ca.pem") // required when sslMode is verify_ca or verify_identity, default null, null means has no server CA cert
    .option(Option.valueOf("sslCert"), "/path/to/mysql/client-cert.pem") // optional, default null, null means has no client cert
    .option(Option.valueOf("sslKey"), "/path/to/mysql/client-key.pem") // optional, default null, null means has no client key
    .option(Option.valueOf("sslKeyPassword"), "key-pem-password-in-here") // optional, default null, null means has no password for client key (i.e. "sslKey")
    .option(Option.valueOf("tlsVersion"), "TLSv1.3,TLSv1.2,TLSv1.1") // optional, default is auto-selected by the server
    .option(Option.valueOf("sslHostnameVerifier"), "com.example.demo.MyVerifier") // optional, default is null, null means use standard verifier
    .option(Option.valueOf("sslContextBuilderCustomizer"), "com.example.demo.MyCustomizer") // optional, default is no-op customizer
    .option(Option.valueOf("zeroDate"), "use_null") // optional, default "use_null"
    .option(Option.valueOf("useServerPrepareStatement"), true) // optional, default false
    .option(Option.valueOf("tcpKeepAlive"), true) // optional, default false
    .option(Option.valueOf("tcpNoDelay"), true) // optional, default false
    .option(Option.valueOf("autodetectExtensions"), false) // optional, default false
    .build();
ConnectionFactory connectionFactory = ConnectionFactories.get(options);

// Creating a Mono using Project Reactor
Mono<Connection> connectionMono = Mono.from(connectionFactory.create());
```

Or use unix domain socket like following:

```java
// Minimum configuration for unix domain socket
ConnectionFactoryOptions options = ConnectionFactoryOptions.builder()
    .option(DRIVER, "mysql")
    .option(Option.valueOf("unixSocket"), "/path/to/mysql.sock")
    .option(USER, "root")
    .build();
ConnectionFactory connectionFactory = ConnectionFactories.get(options);

Mono<Connection> connectionMono = Mono.from(connectionFactory.create());
```

### Programmatic Configuration

```java
MySqlConnectionConfiguration configuration = MySqlConnectionConfiguration.builder()
    .host("127.0.0.1")
    .user("root")
    .port(3306) // optional, default 3306
    .password("database-password-in-here") // optional, default null, null means has no password
    .database("r2dbc") // optional, default null, null means not specifying the database
    .serverZoneId(ZoneId.of("Continent/City")) // optional, default null, null means query server time zone when connection init
    .connectTimeout(Duration.ofSeconds(3)) // optional, default null, null means no timeout
    .socketTimeout(Duration.ofSeconds(4)) // optional, default null, null means no timeout
    .sslMode(SslMode.VERIFY_IDENTITY) // optional, default SslMode.PREFERRED
    .sslCa("/path/to/mysql/ca.pem") // required when sslMode is VERIFY_CA or VERIFY_IDENTITY, default null, null means has no server CA cert
    .sslCert("/path/to/mysql/client-cert.pem") // optional, default has no client SSL certificate
    .sslKey("/path/to/mysql/client-key.pem") // optional, default has no client SSL key
    .sslKeyPassword("key-pem-password-in-here") // optional, default has no client SSL key password
    .tlsVersion(TlsVersions.TLS1_3, TlsVersions.TLS1_2, TlsVersions.TLS1_1) // optional, default is auto-selected by the server
    .sslHostnameVerifier(MyVerifier.INSTANCE) // optional, default is null, null means use standard verifier
    .sslContextBuilderCustomizer(MyCustomizer.INSTANCE) // optional, default is no-op customizer
    .zeroDateOption(ZeroDateOption.USE_NULL) // optional, default ZeroDateOption.USE_NULL
    .useServerPrepareStatement() // Use server-preparing statements, default use client-preparing statements
    .tcpKeepAlive(true) // optional, controls TCP Keep Alive, default is false
    .tcpNoDelay(true) // optional, controls TCP No Delay, default is false
    .autodetectExtensions(false) // optional, controls extension auto-detect, default is true
    .extendWith(MyExtension.INSTANCE) // optional, manual extend an extension into extensions, default using auto-detect
    .build();
ConnectionFactory connectionFactory = MySqlConnectionFactory.from(configuration);

// Creating a Mono using Project Reactor
Mono<Connection> connectionMono = Mono.from(connectionFactory.create());
```

Or use unix domain socket like following:

```java
// Minimum configuration for unix domain socket
MySqlConnectionConfiguration configuration = MySqlConnectionConfiguration.builder()
    .unixSocket("/path/to/mysql.sock")
    .user("root")
    .build();
ConnectionFactory connectionFactory = MySqlConnectionFactory.from(configuration);

Mono<Connection> connectionMono = Mono.from(connectionFactory.create());
```

### Configuration items

| name | valid values | required | description |
|---|---|---|---|
| driver | A constant "mysql" | Required in R2DBC discovery | This driver needs to be discovered by name in R2DBC |
| host | A hostname or IP | Required when `unixSocket` does not exists | The host of MySQL database server |
| unixSocket | An absolute or relative path | Required when `host` does not exists | The `.sock` file of Unix Domain Socket |
| port | A positive integer less than 65536 | Optional, default 3306 | The port of MySQL database server |
| user | A valid MySQL username and not be empty | Required | Who wants to connect to the MySQL database |
| password | Any printable string | Optional, default no password | The password of the MySQL database user |
| database | A valid MySQL database name | Optional, default does not initialize database | Database used by the MySQL connection |
| connectTimeout | A `Duration` which must be positive duration | Optional, default has no timeout | TCP connect timeout |
| socketTimeout | A `Duration` which must be positive duration | Optional, default has no timeout | TCP socket timeout |
| serverZoneId | An id of `ZoneId` | Optional, default query time zone when connection init | Server time zone id |
| tcpKeepAlive | `true` or `false` | Optional, default disabled | Controls TCP KeepAlive |
| tcpNoDelay | `true` or `false` | Optional, default disabled | Controls TCP NoDelay |
| sslMode | A value of `SslMode` | Optional, default `PREFERRED` when using hosting connection, `DISABLED` when using Unix Domain Socket | SSL mode, see following notice |
| sslCa | A path of local file which type is `PEM` | Required when `sslMode` is `VERIFY_CA` or `VERIFY_IDENTITY` | The CA cert of MySQL database server |
| sslCert | A path of local file which type is `PEM` | Required when `sslKey` exists | The SSL cert of client |
| sslKey | A path of local file which type is `PEM` | Required when `sslCert` exists | The SSL key of client |
| sslKeyPassword | Any valid password for `PEM` file | Optional, default `sslKey` has no password | The password for client SSL key (i.e. `sslKey`) |
| tlsVersion | Any value list of `TlsVersions` | Optional, default is auto-selected by the server | The TLS version for SSL, see following notice |
| sslHostnameVerifier | A `HostnameVerifier` | Optional, default use RFC standard | Used only if `SslMode` is `VERIFY_CA` or higher |
| sslContextBuilderCustomizer | A `Function<SslContextBuilder, SslContextBuilder>` | Optional, default is NO-OP function | Used only if `SslMode` is not `DISABLED` |
| zeroDateOption | Any value of `ZeroDateOption` | Optional, default `USE_NULL` | The option indicates "zero date" handling, see following notice |
| autodetectExtensions | `true` or `false` | Optional, default is `true` | Controls auto-detect `Extension`s |
| useServerPrepareStatement | `true`, `false` or `Predicate<String>` | Optional, default is `false` | See following notice |

- `SslMode` Considers security level and verification for SSL, make sure the database server supports SSL before you want change SSL mode to `REQUIRED` or higher. **The Unix Domain Socket only offers "DISABLED" available**
  - `DISABLED` I don't care about security and don't want to pay the overhead for encryption
  - `PREFERRED` I don't care about encryption but will pay the overhead of encryption if the server supports it. **Unavailable on Unix Domain Socket**
  - `REQUIRED` I want my data to be encrypted, and I accept the overhead. I trust that the network will make sure I always connect to the server I want. **Unavailable on Unix Domain Socket**
  - `VERIFY_CA` I want my data encrypted, and I accept the overhead. I want to be sure I connect to a server that I trust. **Unavailable on Unix Domain Socket**
  - `VERIFY_IDENTITY` (the highest level, most like web browser): I want my data encrypted, and I accept the overhead. I want to be sure I connect to a server I trust, and that it's the one I specify. **Unavailable on Unix Domain Socket**
  - `TUNNEL` Use SSL tunnel to connect to MySQL, it may be useful for some RDS that's using SSL proxy **Unavailable on Unix Domain Socket**
- `TlsVersions` Considers TLS version names for SSL, can be **multi-values** in the configuration, make sure the database server supports selected TLS versions. Usually sorted from higher to lower. **Unavailable on Unix Domain Socket**
  - `TLS1` (i.e. "TLSv1") Under generic circumstances, MySQL database supports it if database supports SSL
  - `TLS1_1` (i.e. "TLSv1.1") Under generic circumstances, MySQL database supports it if database supports SSL
  - `TLS1_2` (i.e. "TLSv1.2") Supported only in Community Edition `8.0.4` or higher, and Enterprise Edition `5.6.0` or higher
  - `TLS1_3` (i.e. "TLSv1.3") Supported only available as of MySQL `8.0.16` or higher, and requires compiling MySQL using OpenSSL `1.1.1` or higher
- `ZeroDateOption` Considers special handling when MySQL database server returning "zero date" (i.e. `0000-00-00 00:00:00`)
  - `EXCEPTION` Just throw an exception when MySQL database server return "zero date"
  - `USE_NULL` Use `null` when MySQL database server return "zero date"
  - `USE_ROUND` **NOT** RECOMMENDED, only for compatibility. Use "round" date (i.e. `0001-01-01 00:00:00`) when MySQL database server return "zero date"
- Prepare Statement: Considers based on server-preparing or client-preparing, some database server maybe not support server-preparing binary-query, such as Vitess
  - `useClientPrepareStatement()` default preparing mode, use client-preparing text-query for parametrized statements
  - `useServerPrepareStatement()` use server-preparing binary-query for parametrized statements
  - `useServerPrepareStatement(Predicate<String>)` use server-preparing binary-query for parametrized statements, and enforce server-preparing usage for simple query (not parametrized statements). The usage is judged by `Predicate`, it's parameter is the simple SQL statement, enforce server-preparing if return `true`
- `extendWith` Manual extend `Extension`, only available in **programmatic configuration**
  - It is **NOT** RECOMMENDED, enable the `autodetectExtensions` is the best way for extensions
  - The `Extensions` will not remove duplicates, make sure it would be not extended twice or more
  - The auto-detected `Extension`s will not affect manual extends and will not remove duplicates

Should use `enum` in [Programmatic](#programmatic-configuration) configuration that not like discovery configurations, except `TlsVersions` (All elements of `TlsVersions` will be always `String` which is case-sensitive).

### Pooling

See [r2dbc-pool](https://github.com/r2dbc/r2dbc-pool).

### Simple statement

```java
connection.createStatement("INSERT INTO `person` (`first_name`, `last_name`) VALUES ('who', 'how')")
    .execute(); // return a Publisher include one Result
```

### Parametrized statement

```java
connection.createStatement("INSERT INTO `person` (`birth`, `nickname`, `show_name`) VALUES (?, ?name, ?name)")
    .bind(0, LocalDateTime.of(2019, 6, 25, 12, 12, 12))
    .bind("name", "Some one") // Not one-to-one binding, call twice of native index-bindings, or call once of name-bindings.
    .add()
    .bind(0, LocalDateTime.of(2009, 6, 25, 12, 12, 12))
    .bind(1, "My Nickname")
    .bind(2, "Naming show")
    .returnGeneratedValues("generated_id")
    .execute(); // return a Publisher include two Results.
```

- All parameters must be bound before execute, even parameter is `null` (use `bindNull` to bind `null`).
- It will be using client-preparing by default, see `useServerPrepareStatement` in configuration.
- In one-to-one binding, because native MySQL prepared statements use index-based parameters, *index-bindings* will have **better** performance than *name-bindings*.

### Batch statement

```java
connection.createBatch()
    .add("INSERT INTO `person` (`first_name`, `last_name`) VALUES ('who', 'how')")
    .add("UPDATE `earth` SET `count` = `count` + 1 WHERE `id` = 'human'")
    .execute(); // return a Publisher include two Results.
```

> The last `;` will be removed if and only if last statement contains ';', and statement has only whitespace follow the last `;`.

### Transactions

```java
connection.beginTransaction()
    .then(Mono.from(connection.createStatement("INSERT INTO `person` (`first_name`, `last_name`) VALUES ('who', 'how')").execute()))
    .flatMap(Result::getRowsUpdated)
    .thenMany(connection.createStatement("INSERT INTO `person` (`birth`, `nickname`, `show_name`) VALUES (?, ?name, ?name)")
        .bind(0, LocalDateTime.of(2019, 6, 25, 12, 12, 12))
        .bind("name", "Some one")
        .add()
        .bind(0, LocalDateTime.of(2009, 6, 25, 12, 12, 12))
        .bind(1, "My Nickname")
        .bind(2, "Naming show")
        .returnGeneratedValues("generated_id")
        .execute())
    .flatMap(Result::getRowsUpdated)
    .then(connection.commitTransaction());
```

## Data Type Mapping

The default built-in `Codec`s  reference table shows the type mapping between [MySQL][m] and Java data types:

| MySQL Type | Unsigned | Support Data Type |
|---|---|---|
| `INT` | `UNSIGNED` | [**`Long`**][java-Long-ref], [`BigInteger`][java-BigInteger-ref] |
| `INT` | `SIGNED` | [**`Integer`**][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref] |
| `TINYINT` | `UNSIGNED` | [**`Short`**][java-Short-ref], [`Integer`][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref], [`Boolean`][java-Boolean-ref] (Size is 1) |
| `TINYINT` | `SIGNED` | [**`Byte`**][java-Byte-ref], [`Short`][java-Short-ref], [`Integer`][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref], [`Boolean`][java-Boolean-ref] (Size is 1) |
| `SMALLINT` | `UNSIGNED` | [**`Integer`**][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref] |
| `SMALLINT` | `SIGNED` | [**`Short`**][java-Short-ref], [`Integer`][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref] |
| `MEDIUMINT` | `SIGNED/UNSIGNED` | [**`Integer`**][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref] |
| `BIGINT` | `UNSIGNED` | [**`BigInteger`**][java-BigInteger-ref], [`Long`][java-Long-ref] (Not check overflow) |
| `BIGINT` | `SIGNED` | [**`Long`**][java-Long-ref], [`BigInteger`][java-BigInteger-ref] |
| `FLOAT` | `SIGNED` / `UNSIGNED` | [**`Float`**][java-Float-ref], [`BigDecimal`][java-BigDecimal-ref] |
| `DOUBLE` | `SIGNED` / `UNSIGNED` | [**`Double`**][java-Double-ref], [`BigDecimal`][java-BigDecimal-ref]  |
| `DECIMAL` | `SIGNED` / `UNSIGNED` | [**`BigDecimal`**][java-BigDecimal-ref], [`Float`][java-Float-ref] (Size less than 7), [`Double`][java-Double-ref] (Size less than 16) |
| `BIT` | - | [**`ByteBuffer`**][java-ByteBuffer-ref], [`BitSet`][java-BitSet-ref], [`Boolean`][java-Boolean-ref] (Size is 1), `byte[]` |
| `DATETIME` / `TIMESTAMP` | - | [**`LocalDateTime`**][java-LocalDateTime-ref], [`ZonedDateTime`][java-ZonedDateTime-ref], [`OffsetDateTime`][java-OffsetDateTime-ref], [`Instant`][java-Instant-ref] |
| `DATE` | - | [**`LocalDate`**][java-LocalDate-ref] |
| `TIME` | - | [**`LocalTime`**][java-LocalTime-ref], [`Duration`][java-Duration-ref], [`OffsetTime`][java-OffsetTime-ref] |
| `YEAR` | - | [**`Short`**][java-Short-ref], [`Integer`][java-Integer-ref], [`Long`][java-Long-ref], [`BigInteger`][java-BigInteger-ref], [`Year`][java-Year-ref] |
| `VARCHAR` / `NVARCHAR` | - | [**`String`**][java-String-ref] |
| `VARBINARY` | - | [**`ByteBuffer`**][java-ByteBuffer-ref], `Blob`, `byte[]` |
| `CHAR` / `NCHAR` | - | [**`String`**][java-String-ref] |
| `ENUM` | - | [**`String`**][java-String-ref], [`Enum<?>`][java-Enum-ref] |
| `SET` | - | **`String[]`**, [`String`][java-String-ref], [`Set<String>`][java-Set-ref] and [`Set<Enum<?>>`][java-Set-ref] ([`Set<T>`][java-Set-ref] need use [`ParameterizedType`][java-ParameterizedType-ref]) |
| `BLOB`s (`LONGBLOB`, etc.) | - | [**`ByteBuffer`**][java-ByteBuffer-ref], `Blob`, `byte[]` |
| `TEXT`s (`LONGTEXT`, etc.) | - | [**`String`**][java-String-ref], `Clob` |
| `JSON` | - | [**`String`**][java-String-ref], `Clob` |
| `GEOMETRY` | - | **`byte[]`**, `Blob` |

## Add a `Codec`

This is an extension of a highly customized driver behavior of encoding parameter or decoding field data.

Example for an extending `Codec` of JSON based-on Jackson.

First, implement a `Codec`, `ParametrizedCodec`, `MassiveCodec` or `MassiveParametrizedCodec`.

- `Codec` the normal codec
  - Data type is `Class<?>`
  - Data buffer size is less than or equal to `Integer.MAX_VALUE`
- `ParametrizedCodec`
  - Data type is `Class<?>` or `ParametrizedType`
  - Data buffer size is less than or equal to `Integer.MAX_VALUE`
- `MassiveCodec`
  - Data type is `Class<?>`
  - Data buffer size is less than or equal to `UnsignedInteger.MAX_VALUE` (Java does not have unsigned integers, it just only represents the range)
- `MassiveParametrizedCodec`
  - Data type is `Class<?>` or `ParametrizedType`
  - Data buffer size is less than or equal to `UnsignedInteger.MAX_VALUE`
  
Actually, `JSON` can store large json data, and its byte size can be `UnsignedInteger.MAX_VALUE`. However, this is just an example.

```java
public final class JacksonCodec implements Codec<Object> {

    /**
     * JUST for example, should configure it in real applications.
     */
    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final ByteBufAllocator allocator;

    /**
     * Used for encoding/decoding mode, see also registrar in second step.
     */
    private final boolean encoding;

    public JacksonCodec(ByteBufAllocator allocator, boolean encoding) {
        this.allocator = allocator;
        this.encoding = encoding;
    }

    @Override
    public Object decode(ByteBuf value, FieldInformation info, Class<?> target, boolean binary, CodecContext context) {
        // If you ensure server is using UTF-8, you can just use InputStream
        try (Reader r = new InputStreamReader(new ByteBufInputStream(value), CharCollation.fromId(info.getCollationId(), context.getServerVersion()).getCharset())) {
            return MAPPER.readValue(r, target);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public Parameter encode(Object value, CodecContext context) {
        return new JacksonParameter(allocator, value, context);
    }

    @Override
    public boolean canDecode(FieldInformation info, Class<?> target) {

CHANGELOG.md
# R2DBC MySQL Changelog

## 0.9.0

### New features
- Supports r2dbc-spi 0.9.0.RELEASE

### Remove Vulnerabilities
- Updated various out-of-dated dependencies.


src/main/java/io/asyncer/r2dbc/mysql/Capability.java
/*
 * Copyright 2023 asyncer.io projects
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.asyncer.r2dbc.mysql;

/**
 * A capabilities flag bitmap considers to define the session behaviors of the connection.
 * <p>
 * Some flags should not be disabled by driver or always enabled. These flags may have no getter or builder
 * setter, or neither.
 */
public final class Capability {

    /**
     * Can use long password.
     */
    private static final int LONG_PASSWORD = 1;

    /**
     * Use found/touched rows instead of changed rows for affected rows. Should enable it by default.
     */
    private static final int FOUND_ROWS = 2;

    /**
     * Use 2-bytes definition flags of {@code DefinitionMetadataMessage}.
     * <p>
     * Very old servers (before 3.23) will not set this capability flag.
     */
    private static final int LONG_FLAG = 4;

    /**
     * Connect to server with a database.
     */
    private static final int CONNECT_WITH_DB = 8;

    /**
     * Enable it to disallow a statement which use {@code database.table.column} for access other schema
     * data.
     */
    private static final int NO_SCHEMA = 16;

    /**
     * The deflate compression, old compression flag.
     */
    private static final int COMPRESS = 32;

//    private static final int ODBC = 64; // R2DBC driver is not ODBC driver.

    /**
     * Allow to use LOAD DATA [LOCAL] INFILE statement.
     */
    private static final int LOCAL_FILES = 128;

    /**
     * Ignore space between a built-in function name and the subsequent parenthesis.
     * <p>
     * Note: Ignoring spaces may cause ambiguity.
     * <p>
     * See also https://dev.mysql.com/doc/refman/8.0/en/function-resolution.html .
     */
    private static final int IGNORE_SPACE = 256;

    /**
     * The protocol version is 4.1 (instead of 3.20).
     */
    private static final int PROTOCOL_41 = 512;

    /**
     * Use {@code wait_interactive_timeout} instead of {@code wait_timeout} for the server waits for activity
     * on a connection before closing it.
     */
    private static final int INTERACTIVE = 1024;

    /**
     * Enable SSL.
     */
    private static final int SSL = 2048;

//    private static final int IGNORE_SIGPIPE = 4096; // Connector/C only flag.

    /**
     * Allow transactions. All available versions of MySQL server support it.
     */
    private static final int TRANSACTIONS = 8192;

    // Old flag and alias of PROTOCOL_41. It will not be used by any available server version/edition.
//    private static final int RESERVED = 16384;

    /**
     * Allow second part of authentication hashing salt.
     * <p>
     * Old name: RESERVED2.
     * <p>
     * Origin name: SECURE_CONNECTION.
     */
    private static final int SECURE_SALT = 32768;

    /**
     * Allow to send multiple statements in text query and prepare query.
     * <p>
     * Old name: MULTI_QUERIES.
     */
    private static final int MULTI_STATEMENTS = 65536;

    /**
     * Allow to receive multiple results in the response of executing a text query.
     */
    private static final int MULTI_RESULTS = 1 << 17;

    /**
     * Allow to receive multiple results in the response of executing a prepare query.
     */
    private static final int PS_MULTI_RESULTS = 1 << 18;

    /**
     * Supports authentication plugins. Server will send more details (i.e. name) for authentication plugin.
     */
    private static final int PLUGIN_AUTH = 1 << 19;

    /**
     * Connection attributes should be sent.
     */
    private static final int CONNECT_ATTRS = 1 << 20;

    /**
     * Can use var-integer sized bytes to encode client authentication.
     * <p>
     * Origin name: PLUGIN_AUTH_LENENC_CLIENT_DATA.
     */
    private static final int VAR_INT_SIZED_AUTH = 1 << 21;

//    private static final int HANDLE_EXPIRED_PASSWORD = 1 << 22; // Client can handle expired passwords.
//    private static final int SESSION_TRACK = 1 << 23;

    /**
     * The MySQL server marks the EOF message as deprecated and use OK message instead.
     */
    private static final int DEPRECATE_EOF = 1 << 24;

    // Allow the server not to send column metadata in result set,
    // should NEVER enable this option.
//    private static final int OPTIONAL_RESULT_SET_METADATA = 1 << 25;
//    private static final int Z_STD_COMPRESSION = 1 << 26;

    // A reserved flag, used to extend the 32-bits capability bitmap to 64-bits.
    // There is no available MySql server version/edition to support it.
//    private static final int CAPABILITY_EXTENSION = 1 << 29;
//    private static final int SSL_VERIFY_SERVER_CERT = 1 << 30; // Client only flag, use SslMode instead.
//    private static final int REMEMBER_OPTIONS = 1 << 31; // Connector/C only flag.

    private static final int ALL_SUPPORTED = LONG_PASSWORD | FOUND_ROWS | LONG_FLAG | CONNECT_WITH_DB |
        NO_SCHEMA | COMPRESS | LOCAL_FILES | IGNORE_SPACE | PROTOCOL_41 | INTERACTIVE | SSL |
        TRANSACTIONS | SECURE_SALT | MULTI_STATEMENTS | MULTI_RESULTS | PS_MULTI_RESULTS |
        PLUGIN_AUTH | CONNECT_ATTRS | VAR_INT_SIZED_AUTH | DEPRECATE_EOF;

    private final int bitmap;

    /**
     * Checks if the connection will be connected and logon with a database.
     *
     * @return if login with database.
     */
    public boolean isConnectWithDatabase() {
        return (bitmap & CONNECT_WITH_DB) != 0;
    }

    /**
     * Checks if the connection enabled SSL.
     *
     * @return if SSL enabled.
     */
    public boolean isSslEnabled() {
        return (bitmap & SSL) != 0;
    }

    /**
     * Checks if the connection is using protocol 4.1.
     *
     * @return if using protocol 4.1.
     */
    public boolean isProtocol41() {
        return (bitmap & PROTOCOL_41) != 0;
    }

    /**
     * Checks if can use var-integer sized bytes to encode client authentication.
     *
     * @return if can use var-integer sized authentication.
     */
    public boolean isVarIntSizedAuthAllowed() {
        return (bitmap & VAR_INT_SIZED_AUTH) != 0;
    }

    /**
     * Checks if allow authentication plugin type name.
     *
     * @return if allowed.
     */
    public boolean isPluginAuthAllowed() {
        return (bitmap & PLUGIN_AUTH) != 0;
    }

    /**
     * Checks if the connection contains connection attributes.
     *
     * @return if has connection attributes.
     */
    public boolean isConnectionAttributesAllowed() {
        return (bitmap & CONNECT_ATTRS) != 0;
    }

    /**
     * Checks if server supports multiple-statement. i.e. computed statement.
     *
     * @return if server supported.
     */
    public boolean isMultiStatementsAllowed() {
        return (bitmap & MULTI_STATEMENTS) != 0;
    }

    /**
     * Checks if server marks EOF message as deprecated.
     *
     * @return if EOF message was deprecated.
     */
    public boolean isEofDeprecated() {
        return (bitmap & DEPRECATE_EOF) != 0;
    }

    /**
     * Checks if server uses more than 8 bytes of salt.
     *
     * @return if using secure salt.
     */
    public boolean isSaltSecured() {
        return (bitmap & SECURE_SALT) != 0;
    }

    /**
     * Checks if server supports transaction.
     *
     * @return if server supported.
     */
    public boolean isTransactionAllowed() {
        return (bitmap & TRANSACTIONS) != 0;
    }

    /**
     * Get the original bitmap of {@link Capability this}.
     *
     * @return the bitmap.
     */
    public int getBitmap() {
        return bitmap;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof Capability)) {
            return false;
        }

        Capability that = (Capability) o;

        return bitmap == that.bitmap;
    }

    @Override
    public int hashCode() {
        return bitmap;
    }

    @Override
    public String toString() {
        // Do not consider complex output, just use hex.
        return "Capability<0x" + Integer.toHexString(bitmap) + '>';
    }

    Builder mutate() {
        return new Builder(bitmap);
    }

    private Capability(int bitmap) {
        this.bitmap = bitmap;
    }

    /**
     * Creates a {@link Capability} with capabilities bitmap. It will unset all unknown flags.
     *
     * @param capabilities the capabilities bitmap.
     * @return the {@link Capability} without unknown flags.
     */
    public static Capability of(int capabilities) {
        return new Capability(capabilities & ALL_SUPPORTED);
    }

    static final class Builder {

        private int bitmap;

        void disableConnectWithDatabase() {
            this.bitmap &= ~CONNECT_WITH_DB;
        }

        void disableDatabasePinned() {
            this.bitmap &= ~NO_SCHEMA;
        }

        void disableCompression() {
            this.bitmap &= ~COMPRESS;
        }

        void disableLoadDataInfile() {
            this.bitmap &= ~LOCAL_FILES;
        }

        void disableIgnoreAmbiguitySpace() {
            this.bitmap &= ~IGNORE_SPACE;
        }

        void disableInteractiveTimeout() {
            this.bitmap &= ~INTERACTIVE;
        }

        void disableSsl() {
            this.bitmap &= ~SSL;
        }

        void disableConnectAttributes() {
            this.bitmap &= ~CONNECT_ATTRS;
        }

        Capability build() {
            return of(this.bitmap);
        }

        private Builder(int bitmap) {
            this.bitmap = bitmap;
        }
    }
}
